Update esp-hal/README (#1228)
esp-rs/esp-hal: esp-hal/README.md

![Crates.io](https://img.shields.io/crates/l/esp-hal?labelColor=1C2C2E&style=flat-square)
[![Matrix](https://img.shields.io/matrix/esp-rs:matrix.org?label=join%20matrix&labelColor=1C2C2E&color=BEC5C9&logo=matrix&style=flat-square)](https://matrix.to/#/#esp-rs:matrix.org)

### Quickstart

[esp-hal] encompasses the hardware abstraction layer across multiple Espressif devices atop peripheral access crate. Implements a number of the traits defined by [embedded-hal](https://github.com/rust-embedded/embedded-hal) for both, `blocking` and `async` APIs.

> To select a concrete chip, correct chip feature has to be used. The current list of chip features is `esp32`, `esp32c2`, `esp32c3`, `esp32c6`, `esp32h2`, `esp32p4`, `esp32s2`, and `esp32s3`.

Reviewer: ```suggestion
To select a chip, the correct chip feature has to be used. The current list of chip features is `esp32`, `esp32c2`, `esp32c3`, `esp32c6`, `esp32h2`, `esp32p4`, `esp32s2`, and `esp32s3`.
```

> By default, the code uses `blocking` API. If you want to use [Embassy] `async` API, the `async` feature has to be used.

Reviewer: async support is supplemented by embassy, we don't actually need the embassy feature for async to work.

> By default, the code uses `blocking` API. If you want to use [Embassy] `async` API, the `async` feature has to be used.

Reviewer: I would say there is a blocking API available by default here.

Reviewer 2: I'm also not sure why we're `<pre></pre>` formatting all these words like blocking and async?

Author: Removed `` and rephrased, thanks!

> `esp-hal` is closely associated with different crates as well. [esp-lp-hal] is HAL for the low-power RISC-V coprocessors found on the ESP32-C6, ESP32-S2, and ESP32-S3 and is part of the [ESP-HAL] package. Another crates, supporting not only WiFi and BLE that are not part of the [ESP-HAL] package could be found [here](https://github.com/esp-rs/esp-hal?tab=readme-ov-file#ancillary-crates).

Reviewer: Why is esp-hal suddenly `ESP-HAL`?

`no_std` HAL implementations for the peripherals which are common among Espressif devices. Implements a number of the traits defined by [embedded-hal](https://github.com/rust-embedded/embedded-hal).

> [esp-hal] encompasses the hardware abstraction layer across multiple Espressif devices atop peripheral access crate. To select a concrete chip, correct chip feature has to be used. The current list of chip features is `esp32, esp32c2, esp32c3, esp32c6, esp32h2, esp32p4, esp32s2, esp32s3`.

Author: Not sure about this list feature, we have it https://github.com/esp-rs/esp-hal?tab=readme-ov-file#esp-hal-1 as well 🤔

Reviewer: This goes in the code documentation too, so I think this is worth having here.

Reviewer 3: IMHO there is no need to repeat it here

Reviewer 3: ah yeah - the README is also for crates.io 👍

> [esp-hal] encompasses the hardware abstraction layer across multiple Espressif devices atop peripheral access crate. To select a concrete chip, correct chip feature has to be used. The current list of chip features is `esp32, esp32c2, esp32c3, esp32c6, esp32h2, esp32p4, esp32s2, esp32s3`.

Reviewer 2: ```suggestion
[esp-hal] encompasses the hardware abstraction layer across multiple Espressif devices atop peripheral access crate. To select a concrete chip, correct chip feature has to be used. The current list of chip features is `esp32`, `esp32c2`, `esp32c3`, `esp32c6`, `esp32h2`, `esp32p4`, `esp32s2`, and `esp32s3`.
```

Sorry to nitpick, but let's please fix the formatting here.

The package tree is as follows:

> - [devices] holds the information about which `peripherals` are supported by the chip HW, `symbols` created by the maintainers to simplify writing drivers for multiple chips, and `efuse` description.

Reviewer 2: ```suggestion
- [devices] holds the information about which `peripherals` are supported by the chip's hardware, `symbols` created by the maintainers to simplify writing drivers for multiple chips, and `efuse` description.
```

- [ld] contains `linker-scripts` defining the layout of the chip's memory sections, specify the location of various symbols (such as ROM functions, for example).

> - [src] is a place, where _all_ peripheral drivers for main (high-power) core are placed. Each peripheral is separated into it's own source file(s).

Reviewer 2: ```suggestion
- [src] is a place, where _all_ peripheral drivers for main (high-power) core are placed. Each peripheral is separated into its own source file(s).
```

> ### Quickstart

Reviewer: This is how to run the examples, not a quick start. I think we should have both sections, and we should emphasise on using esp-template instead of just that token line below

Author: I added another section, hopefully this does look like quickstart :D

After [configuring] the environment and [tooling] you can build and run our [examples]
- make sure, you are in [examples] directory.

> - use `cargo +$TOOLCHAIN $CHIP_FEATURE --bin=$EXAMPLE_NAME`, for example, to run a `hello_world` example for `ESP32-C3` using `nightly` toolchain: `cargo +nightly esp32c3 --bin=hello_world`

Reviewer: I think we settled on using the xtask to run features, but I might be misremembering

> [!NOTE]
>

> > Please note, that some peripherals miss on some chip HW. Always double-check the [devices] of the chip you want to use or description of example.

Reviewer: Nitpick, just write hardware, not worth the abbreviation imo

Reviewer 4: What does `some peripherals miss` mean? Did you mean to write `some peripherals may be missing`? Chip hardware also sounds a bit redundant, I think chip is enough.

Author: I wanted to note that not every esp chip has the same sets of peripherals - I borrowed Bjoern's suggestion from the comment below.

Reviewer 2: This should be reworded then, that is not clear from the current text IMO.

> > Please note, that some peripherals miss on some chip HW. Always double-check the [devices] of the chip you want to use or description of example.

Reviewer: miss => missing

Reviewer 3: or maybe the sentence could be rephrased a bit ... "something missing" sounds a bit negative

Thinking of something like "Please note not all chips support the same set of peripherals." ?

> > Please note, that some peripherals miss on some chip HW. Always double-check the [devices] of the chip you want to use or description of example.

Reviewer: When using the xtask it will tell you if your chip is not supported, so I don't think we need that last bit if we are going to use the xtask exclusively.
